projects/plugins/jetpack/extensions/blocks/ai-paragraph/edit.js

```diff
@@ -4,222 +4,263 @@ import apiFetch from '@wordpress/api-fetch';
-		return sprintf(
-			/** translators: This will be a prompt to OpenAI to generate a post based on the post title, comma-seperated category names and the last 240 characters of content. */
-			__( "This is a post titled '%1$s', published in categories '%2$s':\n\n … %3$s", 'jetpack' ), // eslint-disable-line @wordpress/i18n-no-collapsible-whitespace
-			title,
-			categoryNames,
-			content
-		);
-	}
+	return (
+		<>
+			<div className="content">
+				<RawHTML>{ content }</RawHTML>
+			</div>
+		</>
+	);
+}
 
-	// No content, only title and categories
-	if ( categoryNames.length ) {
-		return sprintf(
-			/** translators: This will be a prompt to OpenAI to generate a post based on the post title, and comma-seperated category names. */
-			__( "This is a post titled '%1$s', published in categories '%2$s':\n", 'jetpack' ), // eslint-disable-line @wordpress/i18n-no-collapsible-whitespace
-			title,
-			categoryNames
-		);
-	}
+export default function Edit( { attributes, setAttributes, clientId } ) {
+	const [ loadingCompletion, setLoadingCompletion ] = useState( false );
+	const [ loadingCategories, setLoadingCategories ] = useState( false );
+	const [ showAnimation, setShowAnimation ] = useState( false );
+	const [ needsMoreCharacters, setNeedsMoreCharacters ] = useState( 0 );
+	const [ showRetry, setShowRetry ] = useState( false );
+	const [ completionFinished, setCompletionFinished ] = useState( !! attributes.content );
+	const [ errorMessage, setErrorMessage ] = useState( false );
 
-	// Title and content
-	if ( content ) {
-		return sprintf(
-			/** translators: This will be a prompt to OpenAI to generate a post based on the post title, and the last 240 characters of content. */
-			__( "This is a post titled '%1$s':\n\n…%2$s", 'jetpack' ), // eslint-disable-line @wordpress/i18n-no-collapsible-whitespace
-			title,
-			content
-		);
-	}
+	// Let's grab post data so that we can do something smart.
+	const currentPostTitle = useSelect(
+		select => select( 'core/editor' ).getEditedPostAttribute( 'title' ),
+		[]
+	);
 
-	return sprintf(
-		/** translators: This will be a prompt to OpenAI to generate a post based on the post title */
-		__( 'Write content of a post titled "%s"', 'jetpack' ),
-		title
+	let loading = false;
+	const categories = useSelect(
+		select => select( 'core/editor' ).getEditedPostAttribute( 'categories' ),
+		[]
 	);
-}
 
-/**
- * Gather data available in Gutenberg and prepare the best prompt we can come up with.
- *
- * @param {Function} select - function returning Gutenberg data store.
- * @returns {string} - prompt ready to pipe to OpenAI.
- */
-function preparePromptBasedOnEditorState( select ) {
-	const editorContent = formatEditorCanvasContentForCompletion(
-		select( 'core/block-editor' )
-	).slice( -240 );
+	const categoryObjects = useSelect( select => {
+		return categories
+			.map( categoryId => {
+				const category = select( 'core' ).getEntityRecord( 'taxonomy', 'category', categoryId );
```

Comment:
Theoretically we should have this data available on the frontend already as this is loaded in category picker. We only need names corresponding to the IDs and would be sweet to do without an API call.

Reply:
Totally agreed. But I could not find how to do that without an API call..